lib/virt_autotest/utils.pm

```diff
@@ -108,13 +111,27 @@ sub ssh_setup {
 }
 
 sub ssh_copy_id {
-    my $guest           = shift;
+    my ($guest, %args) = @_;
+
+    my $username        = $args{username}        // 'root';
+    my $authorized_keys = $args{authorized_keys} // '.ssh/authorized_keys';
+    my $scp             = $args{scp}             // 0;
     my $mode            = is_sle('=11-sp4')             ? ''                      : '-f';
     my $default_ssh_key = (!(get_var('VIRT_AUTOTEST'))) ? "/root/.ssh/id_rsa.pub" : "/var/testvirt.net/.ssh/id_rsa.pub";
     script_retry "nmap $guest -PN -p ssh | grep open", delay => 15, retry => 12;
     assert_script_run "ssh-keyscan $guest >> ~/.ssh/known_hosts";
-    if (script_run("ssh -o PreferredAuthentications=publickey root\@$guest hostname -f") != 0) {
-        exec_and_insert_password("ssh-copy-id -i $default_ssh_key -o StrictHostKeyChecking=no $mode root\@$guest");
+    if (script_run("ssh -o PreferredAuthentications=publickey -o ControlMaster=no $username\@$guest hostname") != 0) {
+        unless ($scp == 1) {
+            exec_and_insert_password("ssh-copy-id -i $default_ssh_key -o PreferredAuthentications=password,keyboard-interactive -o ControlMaster=no $mode $username\@$guest");
+        } else {
+            exec_and_insert_password("ssh -o PreferredAuthentications=password,keyboard-interactive -o ControlMaster=no $username\@$guest 'mkdir .ssh' || true");
```

Comment:
How about making this a bit more readable like

    my $ssh = "ssh -o PreferredAuthentications=password,keyboard-interactive -o ControlMaster=no $username\@$guest"
    exec_and_insert_password("$ssh 'mkdir .ssh' || true");
    ....

Reply:
I did:
```perl
# Our client key is not authorized, we have to type password with evry command
my $options = "-o PreferredAuthentications=password,keyboard-interactive -o ControlMaster=no";
unless ($scp == 1) {
    exec_and_insert_password("ssh-copy-id $options $mode -i $default_ssh_key $username\@$guest");
} else {
    exec_and_insert_password("ssh $options $username\@$guest 'mkdir .ssh' || true");
    exec_and_insert_password("scp $options $default_ssh_key $username\@$guest:'$authorized_keys'");
    if (script_run("nmap $guest -PN -p ssh -sV | grep Windows") == 0) {
        exec_and_insert_password("ssh $options $username\@$guest 'icacls $authorized_keys /remove \"NT AUTHORITY\\Authenticated Users\"'");
        exec_and_insert_password("ssh $options $username\@$guest 'icacls $authorized_keys /inheritance:r'");
    } else {
        exec_and_insert_password("ssh $options $username\@$guest 'chmod 0700 ~/.ssh/'");
        exec_and_insert_password("ssh $options $username\@$guest 'chmod 0644 ~/.ssh/authorized_keys'");
    }
}
assert_script_run "ssh -o PreferredAuthentications=publickey -o ControlMaster=no $username\@$guest hostname";
```